### ISSUE TITLE
biome incorrectly collapses whitespace on closing paren

### ISSUE THREAD

**Issue by maxpetretta** [biomejs/biome#6392]
for the last week or so, the biome extension in zed has been incorrectly collapsing whitespace, specifically when typing a closing paren. the issue seems to only happen when there is an existing return statement within the function (see video). removing the extension from zed fixes the issue.

note: i am **not** saving in the video after typing the closing paren, so formatting should not be taking place

versions:
```
zed v0.178.5
biome extension v0.1.4
```

to reproduce:
```sh
bun init biome-zed-test
cd ./biome-zed-test
bunx biome init # default config used
zed . # try writing a function
```

https://github.com/user-attachments/assets/a6e0e26a-779a-4277-987b-a0dd92c8ccd2

**Comment by tom-james-watson** [comment]
Are you able to reproduce the issue? If there's any way I can help out here, let me know.

**Comment by chrisvariety** [comment]
FYI, still reproduces on extension v0.2.0 (biome 2.0.0)

**Comment by ematipico** [comment]
So, the issue is that the zed formatter is incorrectly set up. This was flagged by a Zed core contributor last week, and the documentation has been updated. 

Please recreate a reproduction by using the correct configuration: https://biomejs.dev/reference/zed/#formatting

**Comment by chrisvariety** [comment]
Hmm, I wish it was that! Using [this reproduction](https://github.com/maxpetretta/biome-zed-whitespace-closing-paren) (updated to biome 2.0.0), and the exact [reference settings.json](https://biomejs.dev/reference/zed/#formatting), the problem still reproduces:

https://github.com/user-attachments/assets/6f6ab105-ad85-4a0b-96a9-6fe7ea55f14b

**Comment by ematipico** [comment]
Can you reproduce this issue in the playground? The current version of zed isn't compatible with Biome v2. This leads me to think that it's not Biome that is formatting the file.

**Comment by chrisvariety** [comment]
No, it [doesn't reproduce in the Biome Playground](https://biomejs.dev/playground/?code=LwAqACoACgAgACoAIABJAGYAIABwAG8AcwBzAGkAYgBsAGUALAAgAHAAdQB0ACAAeQBvAHUAcgAgAG0AaQBuAGkAbQBhAGwAIAByAGUAcAByAG8AZAB1AGMAdABpAG8AbgAgAGkAbgAgAHQAaABpAHMAIABmAGkAbABlACwAIABiAHUAdAAgAGYAZQBlAGwAIABmAHIAZQBlACAAdABvAAoAIAAqACAAYwByAGUAYQB0AGUAIABvAHQAaABlAHIAIABmAGkAbABlAHMAIABlAGwAcwBlAHcAaABlAHIAZQAgAGkAbgAgAHQAaABlACAAcAByAG8AagBlAGMAdAAgAGkAZgAgAG4AZQBjAGUAcwBzAGEAcgB5AC4ACgAgACoACgAgACoAIABPAG4AYwBlACAAeQBvAHUAIABoAGEAdgBlACAAYQAgAG0AaQBuAGkAbQBhAGwAIAByAGUAcAByAG8AZAB1AGMAdABpAG8AbgAsACAAYwBvAG0AbQBpAHQAIABpAHQAIABhAG4AZAAgAHAAdQBzAGgAIABpAHQAIAB0AG8AIABhACAAcAB1AGIAbABpAGMACgAgACoAIABHAGkAdABIAHUAYgAgAHIAZQBwAG8ALgAKACAAKgAvAAoACgBlAHgAcABvAHIAdAAgAGEAcwB5AG4AYwAgAGYAdQBuAGMAdABpAG8AbgAgAG0AYQBpAG4AKAApACAAewAKAAkALwAvACAARQB4AGkAcwB0AGkAbgBnACAAYwBvAGQAZQAKAAkAYwBvAG4AcwB0ACAAbgB1AG0AIAA9ACAATQBhAHQAaAAuAHIAYQBuAGQAbwBtACgAKQA7AAoACQBpAGYAIAAoAG4AdQBtACAAPgAgADAALgA1ACkAIAB7AAoACQAJAGMAbwBuAHMAbwBsAGUALgBsAG8AZwAoACIARwByAGUAYQB0AGUAcgAgAHQAaABhAG4AIAAwAC4ANQAiACkAOwAKAAkAfQAKAAoACQAvAC8AIABBAGQAZAAgAGEAIABuAGUAdwAgAGMAaABlAGMAawAKAAkALwAvACAAVABPAEQATwA6ACAAdAByAHkAIAByAGUAbQBvAHYAaQBuAGcAIABhAG4AZAAgAHIAZQAtAGEAZABkAGkAbgBnACAAYwBsAG8AcwBpAG4AZwAgAHAAYQByAGUAbgAKAAkAaQBmACAAKABuAHUAbQAgADwAIAAwAC4ANQAgACYAJgAgAG4AdQBtACAAPAAgADAALgA0ACkAIAB7AAoACQAJAGMAbwBuAHMAbwBsAGUALgBsAG8AZwAoACIATABlAHMAcwAgAHQAaABhAG4AIAAwAC4ANQAiACkAOwAKAAkAfQAKAAoACQAvAC8AIABFAHgAaQBzAHQAaQBuAGcAIAByAGUAdAB1AHIAbgAKAAkAcgBlAHQAdQByAG4AIABuAHUAbQA7AAoAfQAKAA%3D%3D).

I can say the whitespace collapse issue is resolved when the biome extension is uninstalled, which would seemingly point at the extension itself, but perhaps there's something else going on there.

Happy to wait for a new Zed version if it's as simple as that. This does reproduce in the latest Zed Preview (0.191.4) as well.

**Comment by siketyan** [comment]
IIRC Zed uses [textDocument/onTypeFormatting](https://microsoft.github.io/language-server-protocol/specifications/lsp/3.17/specification/#textDocument_onTypeFormatting) and it's possible that the Biome LSP responses an incorrect code edits on the request. If it's true, it should be reproducible on VS Code with format on type enabled.

**Comment by chrisvariety** [comment]
Good catch, this does reproduce on VS Code with "Editor: Format On Type" enabled. I guess this should be moved to the main biome repo where the LSP lives? I did a quick search there but didn't find any relevant issues.

https://github.com/user-attachments/assets/a437de9a-1c9d-4fd9-a921-7338b3043e3e

**Comment by ryan-nebulock** [comment]
Just as a note, if you add `"use_on_type_format": false` to your Zed settings, this will avoid the issue. It is not ideal but at least a workaround as the alternative is near un-usable.

**Comment by maelp** [comment]
I have the same issue
